### PR TITLE
Use IFileDialog where available.

### PR DESCRIPTION
**Pull request by ncruces**
Alternative fix for #89.

### CODE REVIEW

**Comment by ncruces** [comment]
Thinking of merging this regardless, wdyt @lonnywong?

**Comment by lonnywong** [comment]
> Thinking of merging this regardless, wdyt @lonnywong?

Sorry for the delay. It doesn't work on that machine. It may fall back to `GetOpenFileName`, I'd like to try debugging to confirm, but he's not free recently.

**Comment by ncruces** [comment]
That's disappointing. If you can help debug, it'd be great. I'll test more just to see if this big change isn't a regression. And I'll probably merge it too.

File: file_windows.go [hunk 1]
```diff
@@ -320,6 +426,22 @@ func browseForFolderCallback(wnd win.HWND, msg uint32, lparam, data uintptr) uin
 	return 0
 }
 
+func coInitialize() (context.CancelFunc, error) {
+	runtime.LockOSThread()
+	err := win.CoInitializeEx(0, win.COINIT_APARTMENTTHREADED|win.COINIT_DISABLE_OLE1DDE)
+	if err == nil || err == win.S_FALSE {
```
Comment (lonnywong) [inline file_windows.go:432]: If the app calls `CoInitializeEx` with `win.COINIT_APARTMENTTHREADED|win.COINIT_DISABLE_OLE1DDE` before calling `zenity`, the `err` will be `win.S_FALSE`. `zenity` should not call `win.CoUninitialize()`.
So, `err == win.S_FALSE` should be moved to below: `err == win.RPC_E_CHANGED_MODE || err == win.S_FALSE`.

Comment (ncruces) [reply file_windows.go:432]: I think we've covered this before:

> To close the COM library gracefully on a thread, each successful call to CoInitialize or CoInitializeEx, **_including any call that returns S_FALSE, must be balanced by a corresponding call to CoUninitialize._**

https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex

Comment (ncruces) [reply file_windows.go:432]: But if you don't want COM to be closed, you can `CoInitializeEx` yourself, and never `CoUninitialize`.
This is a reference count, if you don't `CoUninitialize` it never goes back to zero.

**Comment by lonnywong** [comment]
@ncruces It crashes without falling back to `GetOpenFileName`.
Should we use the solution https://github.com/ncruces/zenity/pull/89/commits/84d5fe5fb3c2a724ce313d340487a01300e1cc8d ? It works with the patch.

**Comment by ncruces** [comment]
If we're going to leave COM initialized, windows can do it for us.

When it returns `S_OK` (`nil`), we leave it initialized.
When it returns `S_FALSE`, it was already initialized, so we can `CoUninitialize` (drop the ref count back to what it was).
When it returns `RPC_E_CHANGED_MODE`, we asked it to change mode, but we don't care about the mode.

Otherwise, it's an error.

**Comment by lonnywong** [comment]
@ncruces Thanks.

https://github.com/ncruces/zenity/pull/89 has been fixed by https://github.com/ncruces/zenity/commit/99f1a258bcfc7eaa8b3a90332e51ae06f8768454: https://github.com/ncruces/zenity/blob/99f1a258bcfc7eaa8b3a90332e51ae06f8768454/file_windows.go#L453-L472

As long as the reference count does not decrease to `0`, the bug will not be triggered.

**Comment by ncruces** [comment]
That was the expectation, great to see it confirmed! It just felt better to leave it to windows to manage the thread local stuff.

**Pull request merged**